### PR TITLE
feat: basic non-aggregates w/ InfluxQL selector functions

### LINKED ISSUE #7533: Teach selector functions how to project additional fields
Teach the selector functions, such as `selector_first`, how to project additional fields for the matching row. An example invocation might be

```sql
SELECT selector_first(usage_idle, time, usage_system, usage_user) FROM cpu
```

and the expected output include a nested `fields` struct with the values of the additional fields for the matching timestamp:

```text
+----------------------------------------------------------------------------------------------------+
| selector_first(t.usage_idle,t.time,t.usage_system,t.usage_user)                                    |
+----------------------------------------------------------------------------------------------------+
| {value: 0.97, time: 1970-01-01T00:00:00.000001, fields: { usage_user: 0.02, usage_system: 0.01 } } |
+----------------------------------------------------------------------------------------------------+
```

> **Note**
>
> The additional field arguments must be column references and not arbitrary expressions.

```[tasklist]
- [x] Teach IOx `selector_` functions how to project additional fields
- [ ] Teach InfluxQL how to use updated IOx `selector_` functions
```

# Rationale

InfluxQL queries which project a single selector function, such as `first`, will produce a matching result and timestamp, such as the following example:

```SQL
SELECT first(usage_idle) FROM cpu
```

```
name: cpu
time                 first
----                 -----
2020-06-11T16:52:00Z 99.8015873015873
```

> **Note**
>
>  The `time` column reflects the timestamp for the `first` value of `usage_idle`, at `2020-06-11T16:52:00Z`

When projecting a single selector, it is possible to include additional fields for the matching timestamp:

```SQL
SELECT first(usage_idle), usage_system, usage_user FROM cpu
```

```
name: cpu
time                 first            usage_system       usage_user
----                 -----            ------------       ----------
2020-06-11T16:52:00Z 99.8015873015873 0.0992063492063492 0.0992063492063492
```

In this case, the columns `usage_system` and `usage_user` project the values of the `first` point at `2020-06-11T16:52:00Z`

### PR DESCRIPTION
**Pull request by crepererum**
I feel this is not the most elegant code but it seems to get stuff done.

~This is currently missing renaming the output if the name clashes, see https://github.com/influxdata/influxdb_iox/issues/7533#issuecomment-1593113915 . I'll do that in a follow-up.~ done in this PR.

Closes #7533.

### CODE REVIEW

File: iox_query_influxql/src/plan/planner.rs [hunk 1]
```diff
@@ -82,6 +82,8 @@ use std::sync::Arc;
 
 use super::ir::DataSourceSchema;
 use super::parse_regex;
+use super::util::contains_expr;
+use super::util_copy::clone_with_replacement;
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:86]: I plan to create an issue in DataFusion to see if we can make the `clone_with_replacement` API `pub`, so we don't have to use this copy.

File: iox_query_influxql/src/plan/planner.rs [hunk 2]
```diff
@@ -821,36 +823,87 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
+                        if (idx == time_column_index) || (idx == selector_index) {
+                            continue;
+                        }
+                        let (expr, out_name) = match expr.clone() {
+                            Expr::Alias(expr, out_name) => (*expr, out_name),
+                            _ => {
+                                return error::internal("other field is not aliased");
+                            }
+                        };
+                        // TODO(marco): what if there's tag called last/first/...?
```
[the remark this answers is not in the document]
Comment (crepererum) [reply iox_query_influxql/src/plan/planner.rs:863]: that the output column of the selector is also called `last`/..., hence the tag must be renamed, see https://github.com/influxdata/influxdb_iox/issues/7533#issuecomment-1593113915 for an example

File: iox_query_influxql/src/plan/planner.rs [hunk 3]
```diff
@@ -821,36 +823,87 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
             // 3. otherwise, project the Unix epoch (0)
             select_exprs[time_column_index] = if let Some(i) = ctx.interval {
                 let stride = lit(ScalarValue::new_interval_mdn(0, 0, i.duration));
-                let offset = i.offset.map_or(0, |v|v);
+                let offset = i.offset.map_or(0, |v| v);
```
[the remark this answers is not in the document]
Comment (crepererum) [reply iox_query_influxql/src/plan/planner.rs]: done

File: iox_query_influxql/src/plan/planner.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner.rs:903]: The `rewriter` handles the renaming automatically, so you should be able to use those aliases safely.

I created a test case with PR #8058 to verify correctness

File: iox_query_influxql/src/plan/planner.rs [hunk 4]
```diff
@@ -860,21 +862,89 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
+                        let (expr, out_name) = match expr.clone() {
+                            Expr::Alias(expr, out_name) => (*expr, out_name),
+                            _ => {
+                                return error::internal("other field is not aliased");
+                            }
+                        };
+                        if group_by_tag_set.contains(&out_name.as_str()) {
+                            continue;
+                        }
+                        // TODO(marco): what if there's tag called last/first/...?
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:903]:
```suggestion
```

File: iox_query_influxql/src/plan/planner.rs [hunk 5]
```diff
@@ -860,21 +862,89 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
-            };
-
             // Take ownership of the alias, so we don't reallocate, and temporarily place a literal
             // `NULL` in its place.
             let Expr::Alias(_, alias) = std::mem::replace(&mut select_exprs[time_column_index], lit(ScalarValue::Null)) else {
                 return error::internal("time column is not an alias")
             };
 
+            // wrap
+
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:947]: Is this an orphaned comment?

File: iox_query_influxql/src/plan/planner.rs [hunk 6]
```diff
@@ -991,11 +1057,11 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
         let select_exprs_post_aggr = select_exprs
             .iter()
-            .zip(fields)
-            .map(|(expr, f)| {
+            .zip(should_fill_expr)
+            .map(|(expr, should_fill)| {
                 // This implements the `FILL(<value>)` strategy, by coalescing any aggregate
                 // expressions to `<value>` when they are `NULL`.
-                let fill_if_null = if fill_if_null.is_some() && is_aggregate_field(f) {
+                let fill_if_null = if fill_if_null.is_some() && should_fill {
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:1064]: 👍🏻

**Review by stuartcarnie — APPROVED** [review]
Looks good, @crepererum! Nice work.

File: iox_query_influxql/src/plan/planner.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (crepererum) [reply iox_query_influxql/src/plan/planner.rs:903]: figured that out during testing, just forgot to remove the comment. thanks :)

File: iox_query_influxql/src/plan/planner.rs (hunk 4 above)
Comment (crepererum) [reply iox_query_influxql/src/plan/planner.rs]: done

File: iox_query_influxql/src/plan/planner.rs (hunk 5 above)
Comment (crepererum) [reply iox_query_influxql/src/plan/planner.rs]: removed

**Pull request merged**